src/utils.jl

```diff
@@ -83,61 +83,81 @@ macro precedence(ps, prec, body)
     end
 end
 
+struct TMP
```

Comment:
I named these guys kinda bad so feel free to rename them. A comment of why it's used (to reduce generated) code might be useful as well.